# QM when no upper bound is known

$$
\phi\sim\mathcal U(0.01, \pi/8)
$$

In [11]:
import numpy as np
from grid_search import evaluate_algorithm, grid_search_algorithm
from brute_force import find_phi_brute_force, find_phi_fixed_budget_brute_force
from linear_search import find_phi_linear_search, find_phi_fixed_budget_linear_search
from binary_search import find_phi_binary_search, find_phi_binary_search_anneal_m, find_phi_fixed_budget_binary_search, find_phi_fixed_budget_binary_search_anneal_m
from reverse_engineering import find_phi_reverse_engineering, find_phi_fixed_budget_reverse_engineering

rng = np.random.default_rng(42)
R = 1000

In [12]:
phi_max = np.pi/16
print(f"{phi_max = }")
phi_min = 0.01

N_min = max(np.pi//(2*phi_max), 1)
N_max = max(np.pi//(2*phi_min), 1)

N_min, N_max

phi_max = 0.19634954084936207


(8.0, 157.0)

In [ ]:
param_grid = {
    "m_exploration": np.logspace(1, 3, num=50, dtype=int),
    "lookback_window": [1, 2, 5],
    "safeguard": [0, 1, 2, 5],
    "budget": [100_000],
    "inc": [1, 2, 5, 10]
}

results = grid_search_algorithm(
    find_phi_fn=find_phi_fixed_budget_linear_search,
    param_grid=param_grid,
    R=R,
    phi_min=phi_min,
    phi_max=phi_max,
    eps=1e-3,
    rng=rng,
    lambda_budget=0,
)

Using 4 parallel workers


Grid search | workers=4: 100%|██████████| 2400/2400 [13:27<00:00,  2.97it/s, loss=1.56e-02, success=0.38]

=== Grid search summary ===
Total runtime          : 13.47 minutes
Total configurations: 2400
Candidates ≥ 0.95: 0

Best by loss:
{'m_exploration': 25, 'lookback_window': 5, 'safeguard': 1, 'budget': 100000, 'inc': 1, 'mean_budget': 100040.654, 'err': 0.00038360727525010694, 'success_rate': 0.947, 'loss': 0.010387672675250106}

No configuration met success threshold.


In [ ]:
param_grid = {
    "m_exploration": np.logspace(1, 3, num=50, dtype=int),
    "safeguard": [0, 1, 2, 5],
    "budget": [100_000],
    "conf": [0.5, 0.8, 0.9]
}

results = grid_search_algorithm(
    find_phi_fn=find_phi_fixed_budget_binary_search,
    param_grid=param_grid,
    R=R,
    phi_min=phi_min,
    phi_max=phi_max,
    eps=1e-3,
    rng=rng,
    lambda_budget=0,
)

Using 4 parallel workers


Grid search | workers=4: 100%|██████████| 600/600 [03:43<00:00,  2.68it/s, loss=9.69e-02, success=0.11]

=== Grid search summary ===
Total runtime          : 3.73 minutes
Total configurations: 600
Candidates ≥ 0.95: 0

Best by loss:
{'m_exploration': 86, 'safeguard': 2, 'budget': 100000, 'conf': 0.5, 'mean_budget': 99993.796, 'err': 0.0005061814790304226, 'success_rate': 0.908, 'loss': 0.010505561079030423}

No configuration met success threshold.


Using 4 parallel workers
Grid search | workers=4: 100%|██████████| 600/600 [03:43<00:00,  2.68it/s, loss=9.69e-02, success=0.11]=== Grid search summary ===
Total runtime          : 3.73 minutes
Total configurations: 600
Candidates ≥ 0.95: 0

Best by loss:
{'m_exploration': 86, 'safeguard': 2, 'budget': 100000, 'conf': 0.5, 'mean_budget': 99993.796, 'err': 0.0005061814790304226, 'success_rate': 0.908, 'loss': 0.010505561079030423}

No configuration met success threshold.

In [ ]:
param_grid = {
    "m_exploration": np.logspace(1, 3, num=50, dtype=int),
    "safeguard": [0, 1, 2, 5],
    "budget": [100_000],
    "conf": [0.5, 0.8, 0.9],
    "max_b_steps_sub": [2, 3, 4],
    "delta": [3, 4, 5],
}

results = grid_search_algorithm(
    find_phi_fn=find_phi_fixed_budget_binary_search_anneal_m,
    param_grid=param_grid,
    R=R,
    phi_min=phi_min,
    phi_max=phi_max,
    eps=1e-3,
    rng=rng,
    lambda_budget=0,
)

Using 4 parallel workers


Grid search | workers=4: 100%|██████████| 5400/5400 [33:17<00:00,  2.70it/s, loss=9.91e-02, success=0.10]


=== Grid search summary ===
Total runtime          : 33.30 minutes
Total configurations: 5400
Candidates ≥ 0.95: 0

Best by loss:
{'m_exploration': 95, 'safeguard': 5, 'budget': 100000, 'conf': 0.5, 'max_b_steps_sub': 3, 'delta': 5, 'mean_budget': 100036.211, 'err': 0.0005114775379201221, 'success_rate': 0.892, 'loss': 0.01051509863792012}

No configuration met success threshold.


In [18]:
param_grid = {
    "m_exploration": np.logspace(1, 3, num=50, dtype=int),
    "safeguard": [0.8, 0.9, 0.95],
    "budget": [100_000],
}

results = grid_search_algorithm(
    find_phi_fn=find_phi_fixed_budget_reverse_engineering,
    param_grid=param_grid,
    R=R,
    phi_min=phi_min,
    phi_max=phi_max,
    eps=1e-3,
    rng=rng,
    lambda_budget=0,
)

Using 4 parallel workers


Grid search | workers=4: 100%|██████████| 150/150 [00:50<00:00,  2.96it/s, loss=4.62e-04, success=0.94]

✓ Updated best_overall.json
=== Grid search summary ===
Total runtime          : 0.85 minutes
Total configurations: 150
Candidates ≥ 0.95: 6

Best by loss:
{'m_exploration': 828, 'safeguard': 0.9, 'budget': 100000, 'mean_budget': 99990.893, 'err': 0.00037081046933015765, 'success_rate': 0.954, 'loss': 0.00037081046933015765}

Best by budget (success-constrained):
{'m_exploration': 568, 'safeguard': 0.9, 'budget': 100000, 'mean_budget': 99989.205, 'err': 0.00039587319960477236, 'success_rate': 0.952, 'loss': 0.00039587319960477236}


# Low budget high range

In [19]:
phi_max = np.pi/2
print(f"{phi_max = }")
phi_min = 0.01

N_min = max(np.pi//(2*phi_max), 1)
N_max = max(np.pi//(2*phi_min), 1)

N_min, N_max

phi_max = 1.5707963267948966


(1.0, 157.0)

In [21]:
param_grid = {
    "m_exploration": np.logspace(1, 3, num=50, dtype=int),
    "lookback_window": [1, 2, 5],
    "safeguard": [0, 1, 2, 5],
    "budget": [10_000],
    "inc": [1, 2, 5, 10]
}

results = grid_search_algorithm(
    find_phi_fn=find_phi_fixed_budget_linear_search,
    param_grid=param_grid,
    R=R,
    phi_min=phi_min,
    phi_max=phi_max,
    eps=1e-3,
    rng=rng,
    lambda_budget=0,
)

Using 4 parallel workers


Grid search | workers=4: 100%|██████████| 2400/2400 [15:29<00:00,  2.58it/s, loss=1.24e-02, success=0.05]

=== Grid search summary ===
Total runtime          : 15.49 minutes
Total configurations: 2400
Candidates ≥ 0.95: 0

Best by loss:
{'m_exploration': 10, 'lookback_window': 2, 'safeguard': 2, 'budget': 10000, 'inc': 1, 'mean_budget': 10002.613, 'err': 0.003498197189815692, 'success_rate': 0.193, 'loss': 0.003498197189815692}

No configuration met success threshold.
